examples/example_parametrized_nodes.py

```diff
@@ -0,0 +1,160 @@
+
+
+# %%
+# We define a target function which contains numerical constants that
+# are not available as constants for the search and need to be found
+# by local search on parameterized nodes.
+
+
+def f_target(x):
+    return np.pi * x[:, 0] + 1.0 + np.pi
+
+
+# %%
+# Then we define a differentiable(!) objective function for the
+# evolution. It accepts a torch tensor as input variable and uses
+# mean-squared error between the expression represented by a given
+# individual and the target function evaluated on a set of random
+# points. This inner objective is then used by actual objective
+# function to determine the fitness of the individual.
+
+
+def inner_objective(f, seed):
+    torch.manual_seed(seed)
+    batch_size = 500
+    x = torch.DoubleTensor(batch_size, 1).uniform_(-5, 5)
+    y = f(x)
+    return torch.nn.MSELoss()(f_target(x), y[:, 0])
+
+
+def objective(individual, seed):
+    if individual.fitness is not None:
+        return individual
+
+    f = individual.to_torch()
+    loss = inner_objective(f, seed)
+    individual.fitness = -loss.item()
+
+    return individual
+
+
+# %%
+# Next, we define the parameters for the population, the genome of
+# individuals, the evolutionary algorithm, and the local search. Note
+# that we add the custom node defined above as a primitive.
+
+population_params = {"n_parents": 1, "mutation_rate": 0.05, "seed": 818821}
+
+genome_params = {
+    "n_inputs": 1,
+    "n_outputs": 1,
+    "n_columns": 5,
+    "n_rows": 1,
+    "levels_back": None,
+    "primitives": (ParametrizedAdd, cgp.Add, cgp.Sub, cgp.Mul),
+}
+
+ea_params = {"n_offsprings": 4, "tournament_size": 1, "n_processes": 2}
+
+evolve_params = {"max_generations": 500, "min_fitness": 0.0}
+
+local_search_params = {"lr": 5e-3, "gradient_steps": 9}
+
+# %%
+# We then create a Population instance and instantiate the local search
+# and evolutionary algorithm.
+
+pop = cgp.Population(**population_params, genome_params=genome_params)
+
+local_search = functools.partial(
+    cgp.local_search.gradient_based,
+    objective=functools.partial(inner_objective, seed=population_params["seed"]),
+    **local_search_params,
+)
+
+ea = cgp.ea.MuPlusLambda(**ea_params, local_search=local_search)
+
+
+# %%
+# We define a recording callback closure for bookkeeping of the progression of the evolution.
+
+history = {}
+history["fitness_champion"] = []
+
+
+def recording_callback(pop):
+    history["fitness_champion"].append(pop.fitness_parents())
+
+
+# %%
+# We fix the seed for the objective function to make sure results are
+# comparable across individuals and, finally, we call the `evolve`
+# method to perform the evolutionary search.
+
+obj = functools.partial(objective, seed=population_params["seed"])
+
+cgp.evolve(
+    pop, obj, ea, **evolve_params, print_progress=True, callback=recording_callback,
+)
+
+# %%
+# After the evolutionary search has ended, we print the expression
+# with the highest fitness and plot the progression of the search.
```

Comment:
Here, I would be interested to see the development of expressions over the course of the evolution. The fitness seems to increase very smoothly. Has the evolution found the right expression right in the beginning and then just keeps finetuning this over the course of the evolution?

Reply:
i've improved the output of the example, this now shows that for the first few generations an expression without constants is used, then the correct form is found and the rest of the generations are busy finding the correct numerical values.